Setup CLI

alueck opened the pull request:
### What is the new feature?
A platform independent `Setup.CLI` project. I created a new `Setup.Core` project which contains all the core features of the setup tool. `Setup.CLI` and `Setup.GUI` are built on top of this. This PR also resolves issues with decompilation on Linux/MacOS systems thanks to @Chicken-Bones. The CI build now also uses the setup CLI.

### Why should this be part of tModLoader?
So people can also contribute to tModLoader if working with Linux or MacOS systems.

### Are there alternative designs?

### Sample usage for the new feature
Execute `./setup-cli.bat` or `sh ./setup-cli.sh` from the root folder of the repository.

Chicken-Bones commented on setup.sh:
```diff
@@ -0,0 +1,32 @@
+    echo "Created symbolic link for CLI. You can now use ./Setup.CLI"
+else
+    echo "Successfully built CLI. You can now use ./setup/bin/Release/net8.0/Setup.CLI"
+fi
```
The `setup.sh` command shouldn't make a symlink. Instead, it should pass all the args to `setup.sh` through to the built `Setup.CLI` executable. 

We want to make sure that people have up to date submodules and setup project whenever they run it. Having a permanent link to a potentially stale `Setup.CLI` circumvents this.

Chicken-Bones replied:
It would also be good to add a `setup_cli.bat` for windows users

alueck replied:
Actually had the same thought. Done. Does not feel as snappy though, because `dotnet build` always runs first.

Chicken-Bones commented on setup/.gitignore:
```diff
@@ -0,0 +1 @@
+appsettings.json
```
I like that you have combined the settings for both GUI and CLI, I was going to suggest this.
I don't like that it's just called `appsettings.json` in the root. Perhaps `setup.user` or `setup.usersettings`

I'd also like to include a migration of the old settings to the new format when the UI project is launched.

alueck replied:
File is named `user.settings` now and settings are migrated.

Chicken-Bones commented on setup/setup.Designer.csproj:
```diff
@@ -1,16 +0,0 @@
-﻿<Project Sdk="Microsoft.NET.Sdk">
```
Removing this means we can't use the VS UI designer to adjust the settings form

alueck replied:
Re-added as Setup.GUI.Designer.csproj

Chicken-Bones commented on setup/Setup.CLI/Program.cs:
```diff
@@ -0,0 +1,34 @@
+app.AddCommands<RegenSourceCommand>();
+app.AddCommands<OtherCommands>();
+
+await app.RunAsync();
```
The app looks really good. Some things I noticed:
- The help text for the commands is too long. It gets line wrapped on the default console with, and the line wrapping makes it hard to read because the rest of the line isn't indented.
- Typing `Setup.CLI patch` still shows `Usage: Setup.CLI [command]` rather than `Usage: Setup.CLI patch [command]`
- The help text includes the line `Setup.CLI` after the `Usage:` and before `Commands:` which just looks a bit ugly and unhelpful, can we remove it?

Can we give it a name like `setup_cli` perhaps? Since it'll be invoked via `setup_cli.bat` or `setup_cli.sh` anyway.

alueck replied:
- I shortened all descriptions
- Seems like a bug in Cocona's help generation. This would work if I did not use the class based style. Will take a look.
- I actually prefer to see the `Setup.CLI` (now `setup-cli`) after `Usage:`. Before `Commands` you would normally see `patch`, so the same problem like before.

alueck replied:
I switched the CLI framework from Cocona to Spectre.Console.CLI. Output now looks a bit different.

Chicken-Bones commented on setup/Setup.CLI/Properties/launchSettings.json:
```diff
@@ -0,0 +1,8 @@
+    "Setup.CLI": {
+      "commandName": "Project",
+      "workingDirectory": "$(ProjectDir)/../.."
+				}
```
Formatter

Chicken-Bones commented on setup/Setup.CLI/LiveConsoleProgress.cs:
```diff
@@ -0,0 +1,41 @@
+		if (update) {
+			ctx.Refresh();
+		}
+	}
```
Fantastic work. It looks really good. Even has `Ctrl+C` support.

It would be good if the `setup` task could output some lines saying that each phase has been completed. The final console state also ends up looking a bit silly. 

```
Information Ready for Setup
Any changes in /src will be lost.

Progress: 2297/2297
Old Src Removed











Done
```

alueck replied:
Yeah, that's a bit of a problem with the `Live` control. It also worked a bit weird in a case where an error message was printed through `IUserPrompt`. The message basically started on the first line of the table that is used for the live display and overlapped it. I can see if I find any way to do something about this.
Still better than printing thousands of lines of text, I think 😄

alueck replied:
I fully refactored the progress handling. Works alot better now. Maybe give it a try and see if I missed anything.

Chicken-Bones commented on setup/Setup.Core/Extensions/ServiceCollectionExtensions.cs:
```diff
@@ -0,0 +1,83 @@
+namespace Terraria.ModLoader.Setup.Core;
+
+[ExcludeFromCodeCoverage]
+public static class ServiceCollectionExtensions
```
I understand that the Cocona library you selected for the CLI relies on dependency injection, and thus you migrated the GUI project to use it as well.

This may require further discussion on Discord, but I'm not 'all in' on dependency injection frameworks like these. I'd like to keep their usage minimal, and stick to using regular C# to construct objects where possible.

Can we replace all the factory delegates with regular `new XTask(services.Get<>(), ...)` calls? Even ignoring the expression building that enables it, the need to make and store `.Factory` delegates is a fair bit of additional boilerplate and indirection

Chicken-Bones commented on setup/Setup.CLI/Commands/SetupCommands.cs:
```diff
@@ -0,0 +1,77 @@
+
+		return setupTaskFactory(
+			DecompileTaskParameters.CreateDefault(noPrompts: parameters.NoPrompts, maxDegreeOfParallelism: disableParallelDecompilation ? 1 : null),
+			parameters.NoPrompts);
```
Is there a way we can configure `NoPrompts` at a global level? Passing it through the constructors is a bit of a pain.

alueck replied:
I put it into `ProgramSettings`. It kind of fits there, but will not be persisted.

Chicken-Bones commented on setup/Setup.Core/TerrariaExecutableSetter.cs:
```diff
@@ -0,0 +1,110 @@
+		}
+
+		if (noPrompts) {
+			Console.WriteLine($"Critical failure. Terraria steam directory '{this.programSettings.TerrariaSteamDir}' does not exist.");
+			Environment.Exit(1);
```
Prefer to throw exceptions and let the top level handler print it out.

alueck replied:
Done

Chicken-Bones commented on setup/Setup.CLI/Commands/BaseCommand.cs:
```diff
@@ -0,0 +1,76 @@
+			AnsiConsole.WriteException(exception, ExceptionFormats.NoStackTrace);
+
+			SetupOperation.CreateDirectory(ProgramSettings.LogsDir);
+			await File.WriteAllTextAsync(errorLogFile, exception.ToString(), cancellationToken);
```
Need to print a message which directs people to log file path. (We should do this in the GUI message box too)

alueck replied:
Added to both

Chicken-Bones commented on setup/Setup.Core/PatchTask.cs:
```diff
@@ -0,0 +1,198 @@
+			return;
+		}
+
+		this.userPrompt.Prompt(
+			"Patch Results",
+			$"Patches applied with {failures} failures and {warnings} warnings.\nSee /logs/patch.log for details",
+			Failed() ? PromptSeverity.Error : PromptSeverity.Warning);
```
Perhaps this should be called `Inform` rather than `Prompt`? Just so it's clear that it still runs even when `NoPrompts` is set

alueck replied:
Renamed. I struggled a bit with the naming there.

Chicken-Bones commented on setup/Setup.CLI/Commands/PatchSubCommands.cs:
```diff
@@ -0,0 +1,62 @@
+{
+	[Option('m', Description = "Set the patch mode. Default is EXACT. If you change the patch mode, it will be saved and can be omitted for further calls.")]
+	[HasDefaultValue]
+	public Patcher.Mode? PatchMode { get; set; }
```
I know the default is for this flag to be sticky in GUI, but it's likely that CLI users will not read this explanation in the help text, and that having an argument to a command be 'sticky' will contradict their intuitions. 

If we do want a way to apply this setting permanently via CLI, I recommend either a separate command, or a separate argument (eg `--set-patch-mode`)

alueck replied:
Changed. The save calls were more or less based on a previous state of the application where I always had to save the configuration.

Chicken-Bones commented on setup/Setup.CLI/Commands/SetupCommands.cs:
```diff
@@ -0,0 +1,77 @@
+			}
+		}
+
+		ProgramSettings.Save();
```
Is it wise to save settings when using `setupAuto`? I know CI pipelines normally clean the entire workspace and do a fresh checkout each time, but for reliable testing I think that `setupAuto` shouldn't read from or write to the settings file at all. It should rely entirely on arguments

Chicken-Bones reviewed (changes requested):
Excellent work, it's a large task so it's great to see you've really taken it on fully

alueck committed:
Implement setup CLI

# Conflicts:
#	setup.bat
#	setup/Core/CompositeTask.cs
#	setup/Core/DecompileTask.cs
#	setup/Core/DiffTask.cs
#	setup/Core/FormatTask.cs
#	setup/Core/Formatting/AddVisualNewlinesRewriter.cs
#	setup/Core/Formatting/CollectionInitializerFormatter.cs
#	setup/Core/Formatting/FileScopedNamespaceRewriter.cs
#	setup/Core/Formatting/NoNewlineBetweenFieldsRewriter.cs
#	setup/Core/Formatting/RemoveBracesFromSingleStatementRewriter.cs
#	setup/Core/Formatting/SyntaxUtils.cs
#	setup/Core/Formatting/UnindentRewriter.cs
#	setup/Core/HookGenTask.cs
#	setup/Core/ITaskInterface.cs
#	setup/Core/PatchTask.cs
#	setup/Core/ProgramSetting.cs
#	setup/Core/Ref.cs
#	setup/Core/RegenSourceTask.cs
#	setup/Core/RoslynTask.cs
#	setup/Core/SetupOperation.cs
#	setup/Core/SetupTask.cs
#	setup/Core/SimplifierTask.cs
#	setup/Core/UpdateLocalizationFilesTask.cs
#	setup/Core/Utilities/SteamUtils.cs
#	setup/GUI/AutoSetup.cs
#	setup/GUI/MainForm.cs
#	setup/GUI/Program.cs
#	setup/GUI/Properties/Settings.Designer.cs
#	setup/GUI/Properties/Settings.settings
#	setup/GUI/Properties/launchSettings.json
#	setup/GUI/Setup.Auto.csproj
#	setup/GUI/Setup.GUI.Designer.csproj
#	setup/GUI/Setup.GUI.csproj
#	setup/GUI/app.config
#	setup/setup.Common.targets
#	setup/setup.sln

alueck committed:
Remove obsolete call

alueck committed:
Keep block-scoped namespaces for existing files for easier diffing

alueck committed:
Use tabs in Setup.GUI.Designer.csproj

alueck committed:
Update Setup.GUI.Designer.csproj

alueck committed:
Remove obsolete code

Chicken-Bones committed:
Cleanup diffs a bit

Chicken-Bones commented on setup/GUI/MainForm.resx:
```diff
@@ -60,6 +120,9 @@
   </metadata>
+  <metadata name="toolTipButtons.TrayLocation" type="System.Drawing.Point, System.Drawing, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a">
+    <value>17, 17</value>
+  </metadata>
```
Duplicated property somehow? Does this file need changing at all?

alueck replied:
I reset the file. Visual Studio just kept changing it.

Chicken-Bones commented on setup/GUI/MainForm.cs:
```diff
@@ -1,48 +1,61 @@
 
-			SetPatchMode(Settings.Default.PatchMode);
-			formatDecompiledOutputToolStripMenuItem.Checked = Settings.Default.FormatAfterDecompiling;
+			SetPatchMode(this.programSettings.PatchMode);
```
General request to remove redundant `this.` references across the codebase

Chicken-Bones commented on setup/Core/ProgramSetting.cs:
```diff
@@ -1,22 +1,39 @@
+		public void Set(T value)
+		{
+			setter(programSettings, value);
+			programSettings.Save();
```
Might be nice to only save on change.

Chicken-Bones commented on setup/GUI/MainForm.cs:
```diff
@@ -122,29 +110,30 @@ private void buttonTask_Click(object sender, EventArgs e)
 		}
 
-		private void RunTaskThread(SetupOperation task)
+		private async Task RunTaskThread(SetupOperation task)
```
No point returning `Task` if you're not going to await it and it can't throw. `async void` is fine

Chicken-Bones commented on setup/GUI/MainForm.cs:
```diff
@@ -55,64 +68,39 @@ public MainForm()
 		private void menuItemTerraria_Click(object sender, EventArgs e)
 		{
-			SelectAndSetTerrariaDirectoryDialog();
+			terrariaExecutableSetter.SelectAndSetTerrariaDirectory();
```
Make method `async` and `await` this one to silence the code warning

Chicken-Bones committed:
Disable nullability in HookGenerator

Chicken-Bones committed:
Add SupportedOSPlatform to steam registry key queries

Chicken-Bones commented on setup/Core/UpdateLocalizationFilesTask.cs:
```diff
@@ -1,38 +1,33 @@
-		p.WaitForExit();
-		// MessageBox.Show("Success. Make sure you diff tModLoader after this");
+
+		if (p != null) {
```
Should throw if null

Chicken-Bones commented on setup/Core/UpdateLocalizationFilesTask.cs:
```diff
@@ -1,38 +1,33 @@
+			WorkingDirectory = new FileInfo(UpdateLocalizationFilesPath).Directory!.FullName,
 		});
-		p.WaitForExit();
-		// MessageBox.Show("Success. Make sure you diff tModLoader after this");
```
We need to keep this info message/popup. `userPrompt.Inform`?

Chicken-Bones replied:
I see the console output already has `Make sure to run Diff.` but that comes from python. Maybe we should remove the note from the python script. The message here is better.

alueck replied:
Updated.

Chicken-Bones commented on setup/Core/UpdateLocalizationFilesTask.cs:
```diff
@@ -1,38 +1,33 @@
+	public override async Task Run(IProgress progress, CancellationToken cancellationToken = default)
 	{
-	}
+		using var taskProgress = progress.StartTask("Updating localization files...");
```
This task bugs out the console if run without `--plain-progress`

alueck replied:
Yeah, if anything writes to stdout not using `ITaskProgress` while the live display is rendered, it will bug out. There is not much I can do about that. Maybe we should just disable live progress for this task.

Chicken-Bones commented on setup/CLI/Commands/DiffSubCommands.cs:
```diff
@@ -0,0 +1,72 @@
+		DiffTaskParameters.ForTerraria(programSettings);
+
+	protected override void ResetTimestamp(ProgramSettings programSettings) =>
+		programSettings.TerrariaDiffCutoff = null;
```
I believe you can generify this in the base class via `DiffTaskParameters.Cutoff`

alueck replied:
Yes, thanks. Didn't notice that.

Chicken-Bones committed:
Use dotnet run for faster and cleaner CLI launching

Chicken-Bones committed:
Add a marker file to skip unnecessary slow git submodule restore

Chicken-Bones committed:
Remove unnecessary pause on error in setup-cli.bat

Chicken-Bones committed:
Fix command existence checks in setup-cli.sh

Chicken-Bones commented on setup/Core/RoslynTask.cs:
```diff
@@ -1,86 +1,76 @@
-				Console.WriteLine($"Loading project '{projectPath}'");
+		public override async ValueTask<bool> ConfigurationPrompt(CancellationToken cancellationToken = default)
+		{
+			ProjectPath = await projectSelectionPrompt.Prompt(ProjectPath, cancellationToken).ConfigureAwait(false);
```
Would be nice if this could be a CLI arg instead of needing the prompt

alueck replied:
It is now.

Chicken-Bones commented on setup/CLI/CSharpProjectSelectionPrompt.cs:
```diff
@@ -0,0 +1,29 @@
+
+namespace Terraria.ModLoader.Setup.CLI;
+
+public class CSharpProjectSelectionPrompt : ICSharpProjectSelectionPrompt
```
Maybe we don't need this at all if we have cli args for it?

alueck replied:
Yes, deleted.

Chicken-Bones commented on setup/Core/TerrariaExecutableSetter.cs:
```diff
@@ -0,0 +1,109 @@
+			bool continueSelection = userPrompt.Prompt("Error", messageText, PromptOptions.OKCancel, PromptSeverity.Error);
+			if (continueSelection) {
+				terrariaFolderPath = await this.TrySelectTerrariaDirectory(cancellationToken).ConfigureAwait(false);
+			}
```
This should just be informative. The double prompt feels bad in CLI.

 The user can cancel inside the `TrySelectTerrariaDirectory` prompting process

Chicken-Bones replied:
There should also be a way to supply the terraria steam dir from a CLI arg.

Chicken-Bones commented on setup/CLI/UserPrompt.cs:
```diff
@@ -0,0 +1,49 @@
+		this.Inform(caption, text, severity);
+
+		string answer = AnsiConsole.Prompt(
+			new SelectionPrompt<string>().AddChoices(Cancel, options == PromptOptions.OKCancel ? OK : Retry));
```
I think the default selected option for `OK/Cancel` prompts should be `OK`

Chicken-Bones commented on setup/Core/ProgramSettings.cs:
```diff
@@ -0,0 +1,62 @@
+		this.jsonFilePath = jsonFilePath;
+	}
+
+	public string LogsDir { get; set; } = Path.Combine("setup", "logs");
```
No need for this to be a setting. Can just be a `static readonly` or something

Chicken-Bones commented on setup.bat:
```diff
@@ -22,23 +22,23 @@ if NOT ["%errorlevel%"]==["0"] (
 
-	"setup/bin/Debug/net6.0-windows/setup-auto.exe" %2
+	"setup/bin/Release/net8.0/setup-cli.exe" "setup-auto" %2 "steam_build" "--plain-progress"
 ) Else (
```
Please remove the `auto` code from `setup.bat` and instead adjust the `build.yml` to pass the right args to `setup-cli.bat`

alueck replied:
Done

Chicken-Bones commented on setup/Core/SetupOperation.cs:
```diff
@@ -1,194 +1,207 @@
+						CancellationToken = cancellationToken,
+					},
+					async (item, ct) => {
+						ct.ThrowIfCancellationRequested();
```
Do you need this when using `ForEachAsync` with a `CancellationToken`?

alueck replied:
Nope, removed.

Chicken-Bones commented on setup/Core/Ref.cs:
```diff
@@ -1,10 +1,11 @@
-﻿namespace Terraria.ModLoader.Setup
+﻿namespace Terraria.ModLoader.Setup.Core
 {
 	public class Ref<T>
```
May as well make this an inner class of the `SetupOperation`. Alternatively we could probably just make the `List<Ref<string>>` in `ExecuteParallel` into a `List<Func<string>>` and take advantage of closures to simplify things greatly.

alueck replied:
I refactored the List to be `List<WorkItem` and deleted the `Ref` class.

Chicken-Bones commented on setup/Core/Utilities/PathUtils.cs:
```diff
@@ -0,0 +1,21 @@
+{
+	public static class PathUtils
+	{
+		public static string GetCrossPlatformFullPath(string path)
```
Can you explain more about this one to me?

alueck replied:
.NET does not correctly resolve the Home directory shortcut `~` by itself. So I had to manually write code to account for this. Otherwise paths input by the user might not work although the path exists. If you use `Path.GetFullPath("~")` you get the current working directory combined with `~` at the end.

Chicken-Bones commented on setup/Core/Utilities/PathUtils.cs:
```diff
@@ -0,0 +1,21 @@
+			return Path.GetFullPath(result);
+		}
+
+		public static string SetCrossPlatformDirectorySeparators(string path) => path.Replace('\\', '/');
```
Perhaps a better name would be `WithUnixSeparators`, since it's called statically under `PathUtils`

Chicken-Bones replied:
It might also be worth making `PathUtils.Combine` or `UnixCombine` to save a bit of verbosity on teh consumer side where we use `Path.Combine` a lot.

alueck replied:
I thought about a `Combine` method. But `Path.Combine` does a bit more than just joining the paths. I added a `UnixJoin` method.

Chicken-Bones committed:
More diff reductions

Chicken-Bones commented on setup/Core/RoslynTask.cs:
```diff
@@ -1,86 +1,68 @@
+			// todo proper error log
+			workspace.WorkspaceFailed += (o, e) => Console.Error.WriteLine(e.Diagnostic.Message);
+
+			Console.WriteLine($"Loading project '{ProjectPath}'");
```
Report/Inform instead of `Console`?

alueck replied:
Yes, thanks.

Chicken-Bones commented on setup/Core/PatchTask.cs:
```diff
@@ -1,175 +1,196 @@
+			DeleteEmptyDirs(parameters.PatchedDir);
 
-			taskInterface.SetStatus("Old Src Removed");
+			taskProgress.ReportStatus("Old Src Removed", overwrite: false);
```
I appreciate the use of named args for boolean constants, but I think it's fine to let the default value do its work here. Other calls in this same file don't pass `overwrite: false`

alueck replied:
Yeah, don't exaclty know why I added it there.

Chicken-Bones commented on setup/Core/PatchTask.cs:
```diff
@@ -1,175 +1,196 @@
+			}
 
-			MessageBox.Show(
+			this.userPrompt.Inform(
```
Unnecessary `this.` please fix usages in solution

Chicken-Bones commented on setup/Core/SetupOperation.cs:
```diff
@@ -23,45 +18,66 @@ public WorkItem(string status, Worker worker) {
 		}
 
-		protected void ExecuteParallel(List<WorkItem> items, bool resetProgress = true, int maxDegree = 0) {
+		private int currentProgress;
```
This can be a local variable

Chicken-Bones commented on setup/Core/PatchTask.cs:
```diff
@@ -1,175 +1,196 @@
-			patcher.Patch(mode);
+			FilePatcher? patcher = FilePatcher.FromPatchFile(patchPath);
+			patcher.Patch(this.programSettings.PatchMode);
 			results.Add(patcher);
```
Consider moving outside this method. Allows us to make `results` local to `Run`

Chicken-Bones commented on setup/Core/PatchTask.cs:
```diff
@@ -1,175 +1,196 @@
+			}
 
-			Log(log.ToString());
+			if (logFile != null) {
```
assert rather than checking

Chicken-Bones commented on setup/Core/SetupOperation.cs:
```diff
@@ -151,44 +156,33 @@ private static bool DeleteEmptyDirsRecursion(string dir) {
-		/// Will cause FinishedDialog to be called if warnings are not supressed
-		/// </summary>
-		/// <returns></returns>
-		public virtual bool Warnings() => false;
```
auto setup (for CI) isn't allowed to complete with warnings. Please account for this.

Chicken-Bones replied:
Perhaps a --strict CLI arg

alueck replied:
The CLI exits with status code 1 now if the task failed or the strict option is passed to `TaskRunner` and the task has warnings. I did not add an additional CLI arg for this, but just passed `strict: true` to `TaskRunner` in the `SetupAutoCommand`.

Chicken-Bones commented on setup/Core/CompositeTask.cs:
```diff
@@ -1,40 +1,52 @@
+				await task.Run(progress, cancellationToken).ConfigureAwait(false);
 				if (task.Failed()) {
 					failed = task;
-					return;
```
`return;` removed? Changes semantics. Maybe it never mattered in practice.

See my other notes about replacing the `Failed` checks with exceptions like normal sane people anyway

Chicken-Bones commented on setup/Core/SetupOperation.cs:
```diff
@@ -151,44 +156,33 @@ private static bool DeleteEmptyDirsRecursion(string dir) {
+		///     Called to display a finished dialog if Failures() || warnings are not supressed and Warnings()
 		/// </summary>
-		public virtual void FinishedDialog() {}
+		public virtual void FinishedPrompt() { }
```
I wonder if we can get rid of `FinishedPrompt` and `Failed` entirely, and just use exceptions and `Inform` in the `Run` methods. 

I can't remember why I had `FinishedPrompt` split out originally. Probably overzealous implementation of `CompositeTask`. I was much greener 9 years ago.

alueck replied:
This is actually required because of what I mentioned here: https://github.com/tModLoader/tModLoader/pull/4353#discussion_r1734974285

Chicken-Bones commented on setup/Core/PatchTask.cs:
```diff
@@ -1,175 +1,196 @@
 
 			var items = new List<WorkItem>();
-			var newFiles = new HashSet<string>();
+			var newFiles = new ConcurrentDictionary<string, object?>();
```
Good spotting.

Chicken-Bones committed:
More diff reductions

Chicken-Bones commented on setup/Core/HookGenTask.cs:
```diff
@@ -1,90 +1,94 @@
-				if (hookType == null || hookILType == null || hookType.IsNested)
+				gen.GenerateFor(type, out TypeDefinition hookType, out TypeDefinition hookILType);
+				if (hookType == null || hookILType == null || hookType.IsNested) {
 					continue;
```
Probably just an overzealous autoformatter, but we do allow single line statements in our code style, and prefer `var`. Update the `.editorconfig` if you need to, but otherwise, try and avoid reformatting code that you aren't changing the semantics of as it makes review and `git blame` harder.

I made some diff reduction commits to help me review earlier

We can do a reformat commit after this PR is merged, with file scoped namespaces and proper brace style and that jazz. Refactor commits are easy to skip when blaming when listed as such.

alueck replied:
Rider doesn't seem to care about the severity of the code styles in the `.editorconfig`. `var style` and braces are configured as `silent` or `suggestion` but were still applied. I might have to change a setting somewhere.

Chicken-Bones commented on setup/Core/Extensions/ServiceCollectionExtensions.cs:
```diff
@@ -0,0 +1,23 @@
+
+		return serviceCollection
+			.AddTransient<TerrariaExecutableSetter>()
+			.AddTransient<TargetsFilesUpdater>()
```
Is the reasoning for `AddTransient` just, prefer `Transient` where possible? Shorter lived objects?

alueck replied:
I used transient when I added the DI framework because many of the existing classes were stateful and just added all the other classes as transient without later reviewing it again. Changed it to `Singleton` because all of these are stateless.

Chicken-Bones commented on .gitignore:
```diff
@@ -36,4 +36,4 @@ solutions/**/*.tar.gz
 solutions/Release/
 **/msbuild.binlog
-/steam_build
+/steam_build
```
accidental removal of newline?

alueck replied:
Well, I even checked out the .gitignore file from 1.4.4 and pushed it, to get rid of this. It still shows up though.

alueck replied:
Nevermind it was just Github unwilling to update the page correctly...

alueck committed:
Remov usage of this

alueck committed:
Only save if settings changed

alueck committed:
Change async-await in MainForm

alueck committed:
Update UpdateLocalizationFilesTask

alueck committed:
Fix warning

alueck committed:
Simplify diff sub commands

alueck committed:
Re-add newline to .gitignore

alueck committed:
Update SetupOperation

alueck committed:
Make UpdateLocalizationFilesTask cross-platform compatible

Process.Start will throw an error if python is not found

alueck committed:
Make OK the default option

alueck committed:
Make LogsDir static

alueck committed:
Make Ref class obsolete

alueck committed:
Rename PathUtils.SetCrossPlatformDirectorySeparators; add UnixJoin method

alueck committed:
Remove obsolete setting of default parameter

alueck committed:
Use progress in RoslynTask

alueck committed:
Make results in PatchTask local

alueck committed:
Use null-forgiving operator for logFile; won't be null

alueck committed:
Break look in CompositeTask if a task failed

alueck committed:
Use singleton instead of transient

alueck committed:
Reset MainForm.resx

alueck committed:
Remove need for CSharpSelectionPrompt

Chicken-Bones committed:
Fix live table renderer bug when status has not been updated yet

Chicken-Bones committed:
Simplify LiveConsoleProgress

Chicken-Bones committed:
Refactor LiveConsoleProgress variable names

Chicken-Bones committed:
Get the localization update task to work with the live progress

Chicken-Bones committed:
Fix inverted time test in setup-cli.sh

Chicken-Bones committed:
Fix Setup.GUI launchSettings

Chicken-Bones committed:
Fix warning

Chicken-Bones committed:
Simplify Terraria selection prompt, and SetupOperation.ConfigurationPrompt using OperationCancelledExceptions

Chicken-Bones committed:
Add new linux Terraria install search path

Chicken-Bones committed:
Revert most of the formatting changes in DecompileTask for better diffs

Chicken-Bones committed:
Remove unnecessary call and clarify comment

alueck commented on setup/CLI/LiveConsoleProgress.cs:
```diff
@@ -45,8 +45,8 @@ private sealed class TaskProgress : ITaskProgress
-		private readonly int row;
-		private readonly int detailsRow;
+		private readonly int HeaderRow;
+		private readonly int StatusRow;
```
@Chicken-Bones These field names don't conform to the naming of any other readonly fields in the solution. Should be lowercase.

alueck replied:
Fixed it.

Chicken-Bones committed:
Fix 'obj\project.assets.json’ doesn’t have a target error when building GUI projects

Chicken-Bones committed:
Cleanup async task running code in MainForm

Chicken-Bones committed:
Set ConfigureAwait(false) to be the default for Setup.Core

Chicken-Bones committed:
Use timer for main form progress updates to avoid bottlenecking or overwhelming on Invoke/BeginInvoke

Chicken-Bones committed:
Add newline to task completion message. Required for unix consoles

Chicken-Bones committed:
Cleanup warnings in DecompileTask

Chicken-Bones committed:
Reference the setup gui build in the output directory

alueck committed:
Cleanup diffs in Formatting folder

alueck committed:
Cleanup HookGenerator diffs

alueck committed:
Use sync method for better readability

alueck committed:
Cleanup some more diffs

alueck committed:
Cleanup

alueck committed:
Only show informative error prompt if Terraria directory cannot be found

alueck committed:
Make commands fail if task failed; make setup-auto command also fail on warnings

alueck committed:
Use setup-cli.bat in build.yml; remove "auto" option from setup.bat

alueck committed:
Fix variable and method names

alueck committed:
Make field readonly

alueck committed:
Return exit code instead of calling `Environment.Exit`

alueck committed:
Fix live progress overflow

alueck committed:
Cleanup

alueck committed:
Remove diff in Setup.GUI.Designer.csproj

alueck committed:
Add missing EnableModelessKeyboardInterop call for PatchReviewer

Chicken-Bones committed:
Fix gui progress bar not animating fast enough

Chicken-Bones committed:
Minor diff minimisations

Chicken-Bones committed:
Use non async work items for a couple of simple file writes

Chicken-Bones committed:
Correctly configure var/explicit type diagnostics.

Chicken-Bones committed:
Remove progress messages from PlainConsoleProgress as they're not helpful

Chicken-Bones committed:
Move encrypted Terraria.exe into the setup program so it can be used for Mac and Linux decomp

Note that the provided Terraria_v1.4.4.9_win.exe.enc has been edited with dnSpy to include SteelSeriesEngineWrapper.dll as an embedded resource for the CI. This will not be required in 1.4.5

alueck committed:
Refactor exception handling

alueck committed:
Cleanup

alueck committed:
Refactor updating of .targets files; add CLI options to set paths

alueck committed:
Also update Git info on every run in GUI project

alueck committed:
Implement Terraria executable retrieval flow for decompilation

alueck committed:
Remove obsolete field from PatchTask

alueck committed:
Refactor AggregateException throwing

alueck committed:
Update docs

alueck committed:
Update comment

alueck committed:
Add strict command line option

alueck committed:
Fix warning

alueck committed:
Remove padding from derived keys

alueck committed:
Remove BaseCommandSettings inheritance for secret commands; base properties are not used

alueck committed:
Add new CLI options to DecompileCommand; update build script

alueck committed:
Remove --no-validate flag from DecompileCommand

alueck committed:
Refactor progress to reduce output for --plain-progress option

alueck committed:
Use other lock object in MainForm

alueck committed:
Use lists to keep WorkItemProgresses to reduce visual noise by items moving around

Chicken-Bones commented on setup/CLI/PlainConsoleProgress.cs:
```diff
@@ -0,0 +1,42 @@
+
+		public void ReportStatus(string status) => WriteStatus(status);
+
+		private void WriteStatus(string status) => Console.WriteLine($"  {status}");
```
If you make `WorkItemProgress` an inner class of `TaskProgress`, then you can access `Indent`. That or just remove the `Indent` method from `TaskProgress` and pull `WriteStatus` out to `PlainConsoleProgress`, make it `static` and use it in both classes

Chicken-Bones commented on .github/workflows/build.yml:
```diff
@@ -213,17 +213,36 @@ jobs:
+          if [ ! -d "./src/decompiled" ]; then
+            ./setup-cli.bat decompile --key $TERRARIA_OWNERSHIP_KEY --terraria-steam-dir ./$TERRARIA_VERSION/Windows --tml-dev-steam-dir steam_build --plain-progress --max-parallelism 1 -f --strict
+
+            err=$?
+            if [ $err -ne 0 ]; then
+              exit $err
+            fi
```
Github actions runs the shell with `-e` so there's no need to handle exit codes yourself.

Chicken-Bones commented on setup/CLI/Commands/SetupAutoCommand.cs:
```diff
@@ -0,0 +1,71 @@
+	public string TMLDevSteamDir { get; set; } = string.Empty;
+}
+
+public sealed class SetupAutoCommand : CancellableAsyncCommand<SetupAutoCommandSettings>
```
Can now be removed

Chicken-Bones commented on setup/Core/SetupOperation.cs:
```diff
@@ -1,96 +1,69 @@
+		protected async Task ExecuteParallel(
+			List<WorkItem> items,
+			ITaskProgress progress,
+			bool resetProgress = true,
```
There are no calls with `resetProgress: false`, so we can remove this parameter.

Chicken-Bones commented on setup/CLI/LiveConsoleProgress.cs:
```diff
@@ -0,0 +1,138 @@
+			lastStatus = status;
+		}
+
+		public IWorkItemProgress StartWorkItem(string status)
+		{
+			lastStatus = string.Empty;
+			var progress = new GenericWorkItemProgress(
+				status,
+				UpdateStatusFromWorkItems,
+				x => {
+					lock (currentWorkItems) {
+						currentWorkItems.Remove(x);
+					}
+				});
+
+			lock (currentWorkItems) {
+				currentWorkItems.Add(progress);
+			}
+
+			UpdateStatusFromWorkItems();
+
+			return progress;
+		}
```
We could probably encapsulate this common code a bit more. The `UpdateStatusFromWorkItems` delegate is the only difference between the `MainForm` and `LiveConsoleProgress` impls

alueck replied:
I added a base class which is now used by `MainForm` and `LiveConsoleProgress`.

alueck committed:
Remove SetupAutoCommand

alueck committed:
Move Indent method in PlainConsoleProgress

alueck committed:
Remove unnecessary exit code checks in build.yml

alueck committed:
Remove unnecessary method parameter

alueck committed:
Add a TaskProgressBase class; use it for GUI and live console progress

alueck committed:
Move WorkItemProgress class into TaskProgressBase

Chicken-Bones commented on setup/CLI/LiveConsoleProgress.cs:
```diff
@@ -0,0 +1,102 @@
+	private readonly LiveDisplayContext context;
+	private readonly Table table;
+	private readonly Timer timer;
+	private readonly ConcurrentDictionary<TaskProgress, object?> taskProgresses = [];
```
Since the way we do tables for live console progress doesn't support two simultaneous progress updating tasks, it's probably better if the code doesn't pretend as such. It should be fine for this to just be `TaskProgress?` and the `onCompleted` just sets `taskProgress = null`. We can include an assertion that `taskProgress == null` in the constructor if we want to...

Alternatively, if we move the `timer` and `Refresh` methods into the `TaskProgress` class, we can `Dispose` it all together, and there's no need to track a reference here. (Similar to how the `MainForm` impl works)

Chicken-Bones committed:
Fix secret commands auto retrieval of Terraria exe path

Chicken-Bones committed:
Fix bash script not passing quoted arguments correctly

Chicken-Bones committed:
Remove embedded SteelSeriesEngineWrapper.dll from Terraria secret. It's now just the regular windows copy

Chicken-Bones committed:
Fix valid culture identification on other platforms

Chicken-Bones committed:
Supply XNA framework dlls and .NET reference assemblies to perfect decompilation on osx/linux

Chicken-Bones committed:
Merge branch '1.4.4' into feature/setup-cli

Chicken-Bones committed:
Remove --max-parallelism 1 from github decompile

Chicken-Bones committed:
Extract zips directly without using temp files

Chicken-Bones committed:
Cleanup

Chicken-Bones committed:
Fix merge

Chicken-Bones commented on setup/Core/TerrariaExecutableSetter.cs:
```diff
@@ -0,0 +1,109 @@
+
+	private void SetTerrariaDirectory(string terrariaSteamDirectory, string? tmlDevSteamDirectoryOverride)
+	{
+		workspaceInfo.UpdatePaths(terrariaSteamDirectory, tmlDevSteamDirectoryOverride ?? workspaceInfo.TMLDevSteamDirectory);
```
This should probably overwrite the tML steam directory each time the Terraria directory is changed. I believe that was the old behavior

alueck replied:
Updated to work that way.

Chicken-Bones committed:
Fix SteamLibraryFoldersRegex for new steam format

Chicken-Bones committed:
Fix merge derp

Chicken-Bones committed:
Make setup-cli.sh executable in git

alueck committed:
Update LiveConsoleProgress

alueck committed:
Add UpdateWorkspaceInfoCommand; fix build.yml; fix file/folder path inputs in CLI

alueck committed:
Inline method

alueck committed:
Fix DecompileCommand description

alueck committed:
Fix docs

alueck committed:
Add constants for paths

alueck committed:
Refactor PathUtils.GetCrossPlatformFullPath

alueck committed:
Also migrate settings on CLI startup

alueck committed:
Fix PathUtils.GetCrossPlatformFullPath

Chicken-Bones committed:
Better cross platform path detection of reference assemblies for HookGen

Chicken-Bones committed:
Improve HookGen progress reporting

Chicken-Bones committed:
Escape markup in status text in LiveConsoleProgress

Chicken-Bones committed:
Cleanup dead code path

Chicken-Bones committed:
Simplify name of TerrariaNetCore patch stage

Chicken-Bones committed:
Use Markup.Escape
